fuse/nodefs/fsconnector.go

```diff
@@ -126,9 +135,23 @@ func (c *FileSystemConnector) forgetUpdate(nodeID uint64, forgetCount int) {
 
 	if forgotten, handled := c.inodeMap.Forget(nodeID, forgetCount); forgotten {
 		node := (*Inode)(unsafe.Pointer(handled))
+		if node.IsDir() && len(node.Children()) > 0 {
+			// This directory node has children that have not yet been forgotten.
+			// We don't want to remove it from the tree (i.e. remove it from its parent)
+			// because the children may have parent pointers to it.
+			// A later LOOKUP will return this node with a new NodeID.
+			return
+		}
 		node.mount.treeLock.Lock()
-		c.recursiveConsiderDropInode(node)
+		// Remove Inode from all parents
```

Comment:
The backlinks needed for inodes are tied to the nodefs.Node implementation,
so they already are in the same structure, at least before your patch
series.

On Wed, Mar 23, 2016 at 9:51 AM rfjakob <notifications@github.com> wrote:

> In fuse/nodefs/fsconnector.go
> <https://github.com/hanwen/go-fuse/pull/103#discussion_r57125477>:
>
> >  		node.mount.treeLock.Lock()
> > -		c.recursiveConsiderDropInode(node)
> > +		// Remove Inode from all parents
>
> It looks like pathfs could just stop using the children map that nodefs
> provides. The lookup already calls into the filesystem if it does not know
> the child, so this would need no code changes in nodefs.
>
> This would get rid of the duplicate tree and allow the hard link tracking
> to live exclusively in pathfs.
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly or view it on GitHub
> <https://github.com/hanwen/go-fuse/pull/103/files/004bb8a47f407fea15ba2caedcb216371ca806f4#r57125477>
>